src/engine.rs

```diff
@@ -191,7 +191,18 @@ where
             debug_assert_eq!(writer.perf_context_diff.write_wait_duration, Duration::ZERO);
             perf_context += &writer.perf_context_diff;
             set_perf_context(perf_context);
-            writer.finish()?
+            // Retry if `writer.finish()` returns a special err, remarking there still
+            // exists free space for this `LogBatch`.
+            let ret = writer.finish();
```

Comment:
This is not enough. fsync failure will panic. You should let leader do the retry.